keps/sig-storage/20190709-csi-snapshot.md

```diff
@@ -0,0 +1,102 @@
+editor: TBD
+creation-date: 2019-07-09
+last-updated: 2019-07-29
+status: implementable
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Title
+
+CSI Snapshot
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Test Plan](#test-plan)
+  - [Unit tests](#unit-tests)
+  - [E2E tests](#e2e-tests)
+- [Graduation Criteria](#graduation-criteria)
+  - [Alpha-&gt;Beta](#alpha-beta)
+  - [Beta-&gt;GA](#beta-ga)
+- [Changes](#changes)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This KEP is written after the original design doc has been approved and implemented. Design for CSI Volume Snapshot Support in Kubernetes is incorporated as part of the [CSI Volume Snapshot in Kubernetes Design Doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-snapshot.md).
+
+The rest of the document includes required information missing from the original design document: test plan and graduation criteria.
+
+## Test Plan
+
+### Unit tests
+
+* Unit tests around snapshot creation and deletion logic.
+* Unit tests around VolumeSnapshot and VolumeSnapshotContent binding logic.
+* Unit tests for creating volume from snapshot.
+
+### E2E tests
+
+* e2e tests for creating/deleting snapshot.
+* e2e tests for creating volume from snapshot.
+
+## Graduation Criteria
+
+### Alpha->Beta
+
+* Feature complete, including:
+  * Create/delete volume snapshots
+  * Create new volumes from a snapshot
+  * SnapshotContent Deletion/Retain Policy
+  * Snapshot Object in Use Protection
+* Unit and e2e tests implemented
+
+### Beta->GA
+
+* Snapshot feature is used as a basic building block in other advanced applications. 
+* Feature deployed in production and have gone through at least one K8s upgrade.
+
+## Changes
+
+Here are the changes since the original design proposal:
+
+* Renamed `Ready` to `ReadyToUse` in the `Status` field of `VolumeSnapshot` API object.
+* Changed type of `RestoreSize` in `CSIVolumeSnapshotSource` from `*resource.Quantity`  to `*int64`.
+* Lease based Leader Election support is added.
+* Added `VolumeSnapshotContent` deletion policy.
+* Added `VolumeSnapshot` and `VolumeSnapshotContent` in Use Protection using Finalizers.
+* Added Finalizer on the snapshot source PVC to prevent it from being deleted when a snapshot is being created from it.
+* Added check to see whether ListSnapshots is supported by the CSI driver. If it is supported, ListSnapshots will be called to find out the status of a snapshot during static binding; otherwise it is assumed the snapshot ID provided by the admin is valid.
+* If snapshot creation times out, we should allow CSI driver to continue to finish the snapshot creation in the background. Timeout will happen at the foreground so VolumeSnapshot is marked as failed. VolumeSnapshotContent will be created but eventually be marked as failed because there is no VolumeSnapshot requesting for it. This work is on-going.
```

Comment:
I think for timeout, VolumeSnapshot status will not be marked as failed so that controller will continue to retry to create until the operation either succeeded or failed.

Reply:
You are right.  Fixed it.